src/libraries/System.Text.RegularExpressions/src/dotnet-regular-expressions.md

```diff
@@ -0,0 +1,237 @@
+# Implementation of System.Text.RegularExpressions
+
+The implementation uses a typical NFA approach that supports back references. Patterns are parsed into a tree (`RegexTree`), translated into an abstract tree (`RegexCode`) by a writer (`RegexWriter`), and then either used in an interpreter (`RegexInterpreter`) or compiled to IL which is executed (`CompiledRegexRunner`). Both of these are instances of `RegexRunner`: in the case of the compiled runner, one must generate a new one from the `RegexCode` using a factory each time the pattern is to be executed.
+
+Regex engines have different features: .NET regular expressions have a couple that others do not have, such as `Capture`s (distinct from `Group`s). It does not support searching UTF-8 text, nor searching a Span over a buffer.
+
+Unlike some DFA based engines, patterns must be trusted. Text may be untrusted with the use of a timeout to prevent catastrophic backtracking.
+
+Performance is important and we welcome optimizations so long as they preserve the public contract.
+
+## Extensibility
+
+Key types have significant protected surface area. This is probably not intended as an general extensibility point, but rather as a detail of implementing saving a compiled regex to disk. Saving to disk is implemented by saving an assembly containing three types, one that derives from each of `Regex`, `RegexRunnerFactory`, and `RegexRunner`. This mechanism accounts for all the protected methods (and even protected fields) on these classes. If we were designing them today, we would likely more carefully limit their public surface, and possibly not rely on derived types.
+
+Protected members are part of the public API which cannot be broken, so they may potentially make some future optimizations more difficult. In particular, we must keep them stable in order to remain compatible with regexes saved by .NET Framework.
```

Comment:
> may potentially make some future optimizations more difficult

And features.  For example, the `string runtext` is exposed as a protected field.  If we wanted to expose the text instead a s a `ReadOnlyMemory<char>`, in order to make it possible to use `Regex` over other sources beyond `string`, it would lead to some challenges.